src/crimson/osd/pg_recovery.cc

```diff
@@ -479,17 +479,6 @@ void PGRecovery::_committed_pushed_object(epoch_t epoch,
   }
 }
 
-template <class EventT>
-void PGRecovery::start_backfill_recovery(const EventT& evt)
-{
-  using BackfillRecovery = crimson::osd::BackfillRecovery;
-  std::ignore = pg->get_shard_services().start_operation<BackfillRecovery>(
-    static_cast<crimson::osd::PG*>(pg),
-    pg->get_shard_services(),
-    pg->get_osdmap_epoch(),
-    evt);
-}
-
```

Comment:
From `BackfillRecovery::do_recovery()`:
```
    // process_event() of our boost::statechart machine is non-reentrant.
    // with the backfill_pipeline we protect it from a second entry from
    // the implementation of BackfillListener.
    // additionally, this stage serves to synchronize with PeeringEvent.
```
Are we ok with not synchronizing backfill events at all?  Perhaps we could instead process `BackfillRecovery::do_recovery()` events concurrently? That way order should be preserved. 
We can introduce a new concurrent phase solely for that and remove the current `peering_pp(*pg).process` stage.

What do you think?

If you disagree, any `BackfillRecovery` reference should be probably removed since it's unused with this commit.

Reply:
> What ordering are you worried about? Different objects may be of radically different sizes and can require different numbers of messages to backfill. The statemachine itself already knows what object recoveries are in flight.
> 
> The only ordering I can think of that we should probably maintain are the messages themselves, and the existing process stage should suffice for that. I think the actual recovery IO can and should be asynchronous.

With this PR we won't actually go through `BackfillRecovery::do_recovery()` so the existing process stage won't be in use. That way, IIUC, we are possibly allowing few concurrent entries to the same state machine.
Moreover, we won't be in sync with other PeeringEvents which contradicts:
``` 
// yes, it's **not** backfilling yet. The PG_STATE_BACKFILLING
  // will be set after on_backfill_reserved() returns.
  // Backfill needs to take this into consideration when scheduling
  // events -- they must be mutually exclusive with PeeringEvent
  // instances. Otherwise the execution might begin without having
  // the state updated.
```

Essentially we are moving `BackfillRecovery` out of the common `BackgroundRecoveryT` logic. As mentioned in the previous comment, any `BackfillRecovery` usage should be also removed so it will be easier to review the impact.

Reply:
> With this PR we won't actually go through `BackfillRecovery::do_recovery()` so the existing process stage won't be in use. That way, IIUC, we are possibly allowing few concurrent entries to the same state machine.

What this PR does is avoid async execution of new BackfilState event created locally, which prevents the locally created event from crossing the pg interval boundary. Outgoing messages still go through `BackfillRecovery::do_recovery()`.

Reply:
> Outgoing messages still go through BackfillRecovery::do_recovery().

`do_recovery()` is _only_ called from `BackgroundRecoveryT<T>::start()`.
By removing the sole `BackfillRecovery` user, which is - `start_operation<BackfillRecovery>`, `BackfillRecovery::do_recovery()` would never be used.
What am I missing here?

Reply:
Ah, yes. You're right. I think `BackfillState` can trigger various backfill related operations asynchronously, so there's no need for another asynchronous machinery, am I right?

Reply:
> Ah, yes. You're right. I think `BackfillState` can trigger various backfill related operations asynchronously, so there's no need for another asynchronous machinery, am I right?

I agree, let's cleanup any `BackfillRecovery` logic then :)
The commit message should state that we are dropping BackfillRecovery since we already have asynchronous machinery in place and there's no need for another one.

Reply:
This can be done in new cleanup commit since

Reply:
Ok, will do:-)

Reply:
I would prefer having the cleanup in this PR and not in a different one so I'll wait with testing by then, is that ok by you?

Reply:
Yeah, totally:-)